File: test/ballot_cipher_test.go
package test

import (
	"encoding/json"
	"math/big"
	"os"
	"testing"

	qt "github.com/frankban/quicktest"
	"github.com/vocdoni/davinci-circom-circuits/test/testutils"
)

func TestBallotCipher(t *testing.T) {
	c := qt.New(t)
	// Get artifact paths
	wasmPath, err := testutils.GetArtifactPath(testutils.BallotCipherWasm)
	c.Assert(err, qt.IsNil)
	zkeyPath, err := testutils.GetArtifactPath(testutils.BallotCipherZkey)
	c.Assert(err, qt.IsNil)
	vkeyPath, err := testutils.GetArtifactPath(testutils.BallotCipherVkey)
	c.Assert(err, qt.IsNil)

	// encrypt ballot
	_, pubKey := testutils.GenerateKeyPair()
	pubX, pubY := testutils.AffineCoords(pubKey)
	k, err := testutils.RandomK()
	c.Assert(err, qt.IsNil, qt.Commentf("Error generating random k"))

	// Circuit derives k_i = Poseidon(k_{i-1}) per field; we only have one field.
	domain := testutils.ToFr(0)
	ks, err := testutils.DerivePoseidonChain(domain, k, 1)
	c.Assert(err, qt.IsNil, qt.Commentf("derive ks"))

	msg := big.NewInt(3)
	c1, c2 := testutils.Encrypt(msg, pubKey, ks[1])
	inputs := map[string]any{
		"encryption_pubkey": []string{pubX.String(), pubY.String()},
		"k":                 k.String(),
		"msg":               msg.String(),
		"c1":                []string{c1[0].String(), c1[1].String()},
		"c2":                []string{c2[0].String(), c2[1].String()},
	}
	bInputs, _ := json.MarshalIndent(inputs, "  ", "  ")
	c.Log("Inputs:", string(bInputs))
	proofData, pubSignals, err := testutils.CompileAndGenerateProof(bInputs, wasmPath, zkeyPath)
	c.Assert(err, qt.IsNil, qt.Commentf("Error compiling and generating proof"))

	// read vkey file
	vkey, err := os.ReadFile(vkeyPath)
	c.Assert(err, qt.IsNil, qt.Commentf("Error reading vkey file"))

	err = testutils.VerifyProof(proofData, pubSignals, vkey)
	c.Assert(err, qt.IsNil, qt.Commentf("Error verifying proof"))
	c.Log("Proof verified")
}

File: test/ts/ballot_cipher_test.ts
import { encrypt, prove, verify } from './utils';

(async () => {
    const wasm = "../../artifacts/ballot_cipher_test.wasm";
    const pk = "../../artifacts/ballot_cipher_test_pkey.zkey";
    const vk = "../../artifacts/ballot_cipher_test_vkey.json";
    // encrypt value
    const pubKey : [bigint, bigint] = [
        BigInt("14683031697277856265190472023105198820413415816394525437006041413571428119286"), 
        BigInt("8231930431069649913978957226360235712312621187451969137400305691913512440493")
    ];
    const k : bigint = BigInt("650538809577380042220943083323036196153738464670981167525900386056100355632");
    const msg : bigint = BigInt("3");
    const [c1, c2] = encrypt(msg, pubKey, k);
    // init inputs
    const inputs = { encryption_pubkey: pubKey, msg, k, c1, c2 };
    console.log("inputs", inputs);
    // generate proof
    const [proof, publicSignals] = await prove(inputs, wasm, pk);
    console.log("proof", proof);
    console.log("pubSignals", publicSignals);
    // verify proof
    const verified = await verify(vk, proof, publicSignals);
    console.log("Proof verified?", verified);
    // exit
    process.exit();
})();